functions.sh

```diff
@@ -293,7 +325,162 @@ mount_by_uuid() {
+	elif ${USE_OPENSSL_RAND:-true}; then
+		openssl rand "${1:-32}" || tpm2 getrandom "${1:-32}"
+	elif [[ -c /dev/urandom ]]; then
+		dd if=/dev/urandom bs="${1:-32}" count=1 2>/dev/null
+	else
+		tpm2 getrandom "${1:-32}" 2>/dev/null || openssl rand "${1:-32}"
+	fi
+}
+
+# Authenticated encryption: confounded AES-256-CBC with HMAC-SHA-256.
+#
+# Confounded means that a block of entropy is prefixed to the plaintext.
+# Kerberos does something similar, but with ciphertext stealing mode (CTS)
+# instead of CBC (CTS is based on CBC).  The use of a confounder means we can
+# use a zero IV, and that we don't need xxd(1) on the decrypt side, possibly
+# making it easier to use aead_decrypt in initramfs.
+#
+# This construction is safe to use with the same key repeatedly, though that is
+# not the intent in Safeboot.dev.
+#
+# OpenSSL really should have a command to do something like this.
+#
+# $1 may be a regular file, a socket, a pipe, /dev/stdin;
+# $2 must be a regular file (seekable);
+# $3 may be a regular file, a socket, a pipe, /dev/stdout.
+aead_encrypt() {
+	local plaintext_file="$1"
+	local key_file="$2"
+	local ciphertext_file="$3"
+	local mackey
+
+	mackey=$(sha256 < "$key_file")
+
+	(_rand 16; cat "$plaintext_file") \
+	| openssl enc -aes-256-cbc					\
+		      -e						\
+		      -nosalt						\
+		      -kfile "$key_file"				\
+		      -iv 00000000000000000000000000000000 2>/dev/null	\
+	|
+	(tee >(openssl dgst -mac HMAC			\
+			     -macopt hexkey:"$mackey"	\
+			     -binary) ) > "$ciphertext_file"
+}
+
+# Authenticated decryption counterpart to aead_encrypt.
+#
+# OpenSSL really should have a command to do something like this.
+#
+# $1 and $2 must be regular files (seekable);
+# $3 may be a regular file or a socket or tty or device.
+#
+# (Making it so $1 can be not-seekable is a pain.  It'd be a lot easier if this
+# was written in Rust, C, or Python.)
+aead_decrypt() {
+	local ciphertext_file="$1"
+	local key_file="$2"
+	local plaintext_file="$3"
+	local mackey sz
+
+	mackey=$(sha256 < "$key_file")
+	sz=$(stat -c '%s' "$ciphertext_file")
+
+	# We add 16 bytes of confounder and 32 bytes of HMAC; OpenSSL will add
+	# some padding
+	((sz >= 48)) || die "ciphertext file too short" || return 1
+
+	# Extract the MAC, compute the MAC as it should be, compare the two
+	# (this complex cmp invocation means we don't need temp files, so no
+	# cleanup either)
+	if cmp <(dd if="$ciphertext_file"				\
+		    iflag=skip_bytes					\
+		    skip=$((sz - 32))					\
+		    bs=32						\
+		    count=1 2>/dev/null)				\
+	       <(dd if="$ciphertext_file"				\
+		    bs=$((sz - 32))					\
+		    count=1 2>/dev/null					\
+		 | openssl dgst -mac HMAC				\
+				-macopt hexkey:"$mackey"		\
+				-binary); then
+		dd if="$ciphertext_file"				\
+		   bs=$((sz - 32))					\
+		   count=1 2>/dev/null					\
+		| openssl enc -aes-256-cbc				\
+			      -d					\
+			      -nosalt					\
+			      -kfile "$key_file"			\
+			      -iv 00000000000000000000000000000000	\
```

Comment:
Are there any concerns about reusing the all-zero IV? I'm not a cryptographer, but I noticed that `tpm2 encryptdecrypt` defaults to IV=0 and also warns about it.

Reply:
Look carefully!  This is "confounded" AES-256-CBC-HMAC-SHA-256.  Which means that a block of nonce bits is prepended to the plaintext, and then that is encrypted in CBC mode with an IV of all-zeros, but the ciphertext block resulting from encrypting the confounder is then the IV for the first real block of plaintext because of the chaining in CBC.  This isn't strictly necessary -- I could have used an explicit IV, and in earlier versions I did, but that complicated the code a fair bit and the IV still had to be included in the output.  Now the code is much simpler and the only cost is an extra block to encrypt/decrypt.

Confounding is used in Kerberos, and when used without additional chaining between messages it is then practically indistinguishable from an explicit IV -- it just costs an extra cipher block operation.  So, I do have an explicit non-zero IV, in a roundabout way.